[2단계 - 장기] 하루(구은선) 미션 제출합니다.

eunseongu opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

안녕하세요 앤지!
1단계 리뷰 꼼꼼하게 해주셔서 감사합니다 👍😊
많은 도움이 되었습니다!
2단계 리뷰도 잘 부탁드립니다!


docker 컨테이너 실행
```bash
docker-compose -p java-janggi up -d
```

docker 컨테이너 종료
```bash
docker-compose -p java-janggi down
```

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
객체지향 생활체조 요구사항을 최대한 지키려고 노력했습니다!
데이터 베이스를 사용하면서 낯선 부분이 많다보니 부족한 부분이 있을 것이라 생각해 3점을 선택했습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

### 1️⃣궁성의 좌표 관리
board에서 궁성의 영역에 들어서면 PalaceArea에서 관련 로직을 처리하도록 시도해보았습니다.
하지만 그렇게 하다 보니 move, validateRoute 등 Board와 PalaceArea 간에 중복되는 로직이 생겼습니다.

그래서 현재는 PalaceArea가 자신의 좌표들만 관리하도록 하여 궁성에 해당하는 위치인지 여부를 판단하는 책임만 가지도록 했습니다.
다만 그러다보니 PalaceArea의 역할이 적은 것 같다는 생각이 들었습니다.
하지만 “궁성에 속한 좌표는 궁성 객체가 아는 것이 자연스럽다”고 생각해 현재 구조를 유지했습니다.

이에 대한 앤지의 생각은 어떠신가요?

---

### 2️⃣`validateRouteObstacles()`, `findRoute()` 메서드
현재는 부모 클래스에 경로 탐색을 위한 `validateRouteObstacles()`, `findRoute()` 메서드가 정의되어 있는데,  
`General`, `Guard`, `Soldier`와 같이 한 칸만 움직이는 기물은 실제로 경로가 존재하지 않아서 빈 구현으로 남겨두고 있습니다.

구현하지 않은 메서드가 있는건 좋은 설계인 것 같지 않아서 고민이 됩니다!
`Routable` 같은 인터페이스를 따로 분리해 경로 개념이 필요한 기물만 해당 기능을 구현하도록 하는 방향이 더 괜찮을까요?

이 구조를 더 명확하게 만들 수 있는 조언 주시면 감사하겠습니다!

---

### 3️⃣현재 구조에서 Dao와 다른 클래스 간의 의존 관계가 적절한지 궁금합니다!

게임을 진행하는 Application에서는 게임의 상태를 관리하기 위해 GameDao를 사용하고 있으며,
이를 통해 게임의 시작, 종료, 차례와 관련된 정보를 다루고 있습니다.

반면, 기물을 관리하는 Board에서는 PieceDao를 사용해
기물의 이동, 조회, 초기화 등 기물 상태와 관련된 로직을 처리하고 있습니다.

이처럼 Dao가 각각의 도메인 역할에 따라 분리되어 사용되도록 구성했는데,
이 구조가 의도한 대로 잘 표현되었는지,
그리고 각 객체가 적절한 의존 관계를 맺고 있다고 보시는지 궁금합니다!

아직 데이터베이스 관련해서는 미숙한 부분이 많아
혹시 개선할 점이나 더 나은 구조가 있다면 어떤 피드백이든 감사히 받겠습니다😊

eunseongu committed:
docs: 들여쓰기 변경

eunseongu committed:
fix: 기물이 없는 위치에 접근할 때 발생하는 예외 처리

- 좌표로 기물을 찾을 때 기물이 존재하는지 확인하지 않아 예외가 발생하는 문제를 해결합니다.

eunseongu committed:
refactor: 게임 진행 로직 변경

eunseongu committed:
refactor: 뷰 리팩토링

- 좌표 입력 방식과 입력 값 처리 방식을 변경합니다.
- 에러메시지를 뷰에서 출력하도록 변경합니다.

eunseongu committed:
feat: 게임 종료 로직 구현

eunseongu committed:
refactor: 기물 정보 객체에서 뷰 의존성 제거

- 기물 객체에서 PieceCategory를 사용하도록 하여 뷰를 의존하지 않도록 변경합니다.

eunseongu committed:
refactor: 장군 판단 로직 변경

eunseongu committed:
docs: 궁성 기능 구현 목록 작성

eunseongu committed:
feat: 장군 모델 구현

eunseongu committed:
feat: 궁성 영역 로직 구현

eunseongu committed:
feat: 대각선 정렬 여부 및 대각선상에 있는 좌표 구하는 기능 추가

eunseongu committed:
feat: 궁성 영역에서 포 움직임 검증 및 경로 찾는 기능 구현

eunseongu committed:
refactor: 진영 판단 로직 수정

- 한인지 초인지 묻는 로직을 같은 진영인지 판단하는 로직으로 변경합니다.

eunseongu committed:
refactor: 움직임 검증 로직 수정

- 메서드 이름을 변경합니다.
- 검증 로직에 맞게 장군 테스트를 수정합니다.

eunseongu committed:
refactor: 궁에서 포 움직임 검증 방식 변경

eunseongu committed:
docs: 구현하지 않아도 되는 내용 삭제

eunseongu committed:
feat: 궁성 영역에서 사 움직임 검증 및 경로 찾는 기능 구현

eunseongu committed:
feat: 궁성 영역에서 차 움직임 검증 및 경로 찾는 기능 구현

eunseongu committed:
feat: 궁성 영역에서 군인 움직임 검증 및 경로 찾는 기능 구현

eunseongu committed:
refactor: 군인 모델 리팩토링

- 중복 로직을 제거합니다.

eunseongu committed:
docs: 구현할 기능 추가

eunseongu committed:
feat: 점수 계산 기능 구현

eunseongu committed:
refactor: 클래스명 변경

eunseongu committed:
docs: 구현할 기능 추가

eunseongu committed:
refactor: 출력 색상 포맷 방식 변경

- PieceSymbol 클래스 대신 PieceType이 기물의 이름을 반환하도록 변경합니다.
- PieceSymbol 클래스를 진영에 따라 색상을 포맷해주는 클래스로 변경합니다.
- 뷰에서 각 진영의 이름을 해당 색상으로 출력하도록 변경합니다.

eunseongu committed:
feat: 데이터베이스 연결 기능 구현

eunseongu committed:
feat: 데이터 추가, 조회 기능 구현

eunseongu committed:
feat: 데이터 수정 기능 구현

eunseongu committed:
feat: 데이터 삭제 기능 구현

eunseongu committed:
feat: 보드판이 데이터베이스를 사용하도록 구현

eunseongu committed:
feat: 새로운 게임인 경우 보드판을 초기화하는 기능 구현

eunseongu committed:
fix: TRUNCATE 쿼리 실행 시 executeQuery → executeUpdate로 수정

eunseongu committed:
refactor: 보드 생성 객체에서 piece 테이블 의존성 제거

eunseongu committed:
feat: 애플리케이션을 재시작하더라도 이전에 하던 장기 게임을 다시 시작하는 기능 구현

eunseongu committed:
refactor: 중복 개선, 미사용 메서드 제거

eunseongu committed:
refactor: 게임 진행 메서드 개선

eunseongu committed:
refactor: board 테이블을 game 테이블로 변경

eunseongu committed:
refactor: 예외 메시지 추가

eunseongu committed:
chore: db/mysql/data 디렉토리를 .gitignore에 추가

soominsohn commented on src/test/java/janggi/dao/PieceDaoTest.java:
```diff
@@ -0,0 +1,54 @@
+
+public class PieceDaoTest {
+
+    private final TestPieceDao testPieceDao = new TestPieceDao();
```
TestPieceDao는 Dao동작을 학습해보고 싶어서 만드신 걸까요?
아니면 PieceDao의 동작을 PieceDaoTest에서 테스트 해보고 싶으셔서 만드신 걸까요?

PieceDaoTest에서는 PieceDao의 동작을 테스트해봐야할 것 같은데, 이렇게 사용하면 TestPieceDao의 동작만 테스트 하는 것 같아요!

어떻게 하면 Dao를 테스트해 볼 수 있을까요?

eunseongu replied:
앤지 의견에 동의합니다! 

PieceDao의 동작을 PieceDaoTest에서 테스트하고 싶어서 TestPieceDao를 만들었습니다. 

TestPieceDao와 PieceDao의 동작이 동일하지만 PieceDao가 작동한다는 보장을 할 수 없습니다.

TestPieceDao와 PieceDao의 차이점은 다른 데이터베이스를 쓰는 것이므로 해당 부분을 외부에서 주입하도록 변경했습니다.

데이터베이스와의 연결 정보를 관리하는 DatabaseConfig를 DatabaseConnector에 주입하고, DatabaseConnector는 Dao를 생성할 때 주입하도록 변경했습니다.

soominsohn replied:
아주 좋습니다! 👍 
DatabaseConfig의 값들을 설정파일로도 관리할 수 있어요!
요 부분은 천천히 학습해보세용 ㅎㅎ!

soominsohn commented on src/main/java/janggi/board/Board.java:
```diff
@@ -46,44 +56,78 @@ private void validateMoveRequest(Point from, Point to) {
 
     public Piece peek(Point point) {
-        if (!placedPieces.containsKey(point)) {
+        if (pieceDao.findByPoint(point) == null) {
```
null을 직접 다루게 되면 NullPointerException이 발생할 위험이 있어요!
NPE는 프로그램에 버그를 유발할 수 있는 아주 무서운 예외입니다.
어떻게 해결해 볼 수 있을까요?

eunseongu replied:
null을 반환하는 대신 예외를 던지거나 optional을 사용할 수 있습니다.

findByPoint 메서드를 사용하는 곳에서 값이 있을 때와 없을 때를 각각 다르게 처리하므로 예외를 던지는 대신 optional을 사용했습니다.

soominsohn commented on src/main/java/janggi/board/Point.java:
```diff
@@ -14,6 +18,19 @@ public Point(int x, int y) {
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("숫자만 입력 가능합니다.");
+        }
+    }
```
버그를 유발할 수 있는 인덱스를 다루는 코드는, 저같은 사람 리뷰어가 위험성을 캐치해 리뷰를 달기 어려울 수 있어요. 😄 
코드 전체적으로 코드래빗이 달아준 리뷰도 잘 살펴보고 반영해보면 좋을 것 같습니다.

soominsohn commented on src/main/java/janggi/dao/PieceDao.java:
```diff
@@ -0,0 +1,134 @@
+        } catch (SQLException e) {
+            throw new RuntimeException("승리한 캠프 조회 중 오류가 발생했습니다.", e);
+        }
+        return null;
```
마지막 코드라인에서 null을 반환해도 될까요?
어떤 상황에서 마지막 코드라인에 도달 할 수 있을까요?

eunseongu replied:
> null을 반환하는 대신 예외를 던지거나 optional을 사용할 수 있습니다.
> 
> findByPoint 메서드를 사용하는 곳에서 값이 있을 때와 없을 때를 각각 다르게 처리하므로 예외를 던지는 대신 optional을 사용했습니다.

반환값을 사용하지 않는 메서드에서는 예외를 던지도록 변경했습니다!

soominsohn commented on src/main/java/janggi/dao/PieceDao.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    public void clearTable() {
+        String query = "TRUNCATE TABLE piece";
```
TRUNCATE에 대해서 학습해주신걸까요~? 💯 
TRUNCATE가 어떻게 동작하는지 저한테도 공유해 주실 수 있을까요? 😄

eunseongu replied:
테이블을 초기화 시키기 위해 데이터를 삭제하고자 했습니다!
Delete는 하나의 행을 삭제하고, Drop은 테이블 자체를 삭제합니다.
Truncate는 테이블의 구조는 유지한 채 내부 모든 데이터를 삭제합니다.

soominsohn commented on src/main/java/janggi/dao/GameDao.java:
```diff
@@ -0,0 +1,93 @@
+            throw new RuntimeException("진행 중인 게임 ID 조회 중 오류가 발생했습니다.", e);
+        }
+        return 0;
+    }
```
토끼 리뷰를 잘 봐주세요 ㅎㅎ try-catch문 이후의 코드라인을 어떻게 처리하면 좋을지 고민해보면 좋을 것 같아요.

soominsohn commented on src/main/java/janggi/Application.java:
```diff
@@ -4,64 +4,93 @@
 
     public static void main(String[] args) {
         View view = new View();
+        GameDao gameDao = new GameDao();
```
> ### 3️⃣현재 구조에서 Dao와 다른 클래스 간의 의존 관계가 적절한지 궁금합니다!
> 
> 게임을 진행하는 Application에서는 게임의 상태를 관리하기 위해 GameDao를 사용하고 있으며,
> 
> 이를 통해 게임의 시작, 종료, 차례와 관련된 정보를 다루고 있습니다.
> 
> 반면, 기물을 관리하는 Board에서는 PieceDao를 사용해
> 
> 기물의 이동, 조회, 초기화 등 기물 상태와 관련된 로직을 처리하고 있습니다.
> 
> 이처럼 Dao가 각각의 도메인 역할에 따라 분리되어 사용되도록 구성했는데,
> 
> 이 구조가 의도한 대로 잘 표현되었는지,
> 
> 그리고 각 객체가 적절한 의존 관계를 맺고 있다고 보시는지 궁금합니다!
> 
> 아직 데이터베이스 관련해서는 미숙한 부분이 많아
> 
> 혹시 개선할 점이나 더 나은 구조가 있다면 어떤 피드백이든 감사히 받겠습니다😊

기물은 PieceDao를 통해, 게임 상태는 GameDao를 통해 조회해오도록 Dao를 분리하신 것은 좋습니다!

- `Board`는 도메인 객체로 보이는데, 도메인이 Dao를 의존해도 괜찮을까요?
- Dao는 어떤 계층이고, 어떤 책임을 위해 등장했을까요?

데이터베이스가 도메인의 상호작용에 영향을 주면 안된다고 생각이들어요. 저라면 아래와 같은 방법으로 작성했을 것 같은데 어떻게 생각하시나요?

1. Dao를 통해 데이터베이스에서 조회한 값을 적절한 도메인으로 변환하고 
2. 도메인끼리의 상호작용으로 장기 턴을 진행 한 후 
3. 그 결과를 다시 Dao를 통해 DB데이터베이스에 저장한다.

이렇게 되면 도메인은 DB와 무관하게 순수하게 유지될 것 같아요.

우선 도메인 객체들을 domain 패키지 하위로 묶어 본 뒤에 현재 구조를 개선할 수 있는 방법에 대해 고민해보면 어떨까요?

eunseongu replied:
앤지의 말씀을 듣고 고민해보니 도메인은 Dao에 의존하면 안된다고 생각합니다. Dao에 의존하면 DB에 의존하게 되는 것이고, DB가 사라지면 동작하지 않는 도메인이 됩니다.

Board가 Dao에 의존하지 않고 도메인끼리 상호작용하도록 하기 위해서 Board 내부에서 map을 이용해 기물들의 좌표를 관리하도록 변경했습니다.

이 과정에서 궁금증이 하나 생겼는데요! 도메인끼리 상호작용하고, Dao를 통해 DB에 저장하도록 하기 위해서 

> Board 내부에서 기물 움직임을 검증하고 움직인 후 좌표를 저장한다 → Dao를 통해 변경된 결과를 저장하거나 수정, 삭제한다.

위 과정을 거치도록 변경했습니다.

그리고 이 과정에서 출발 지점이나 도착 지점에 존재하는 기물을 찾아야 합니다.

이때 Board의 map에서 값을 찾아 반환하도록 하는 것이 좋을까요? 아니면 DB에 저장된 값을 찾아 사용하는 것이 좋을까요?

도메인이 DB에 의존하지 않고 독립적으로 동작하도록 하기 위해서는 peek 메서드를 유지하여 map에서 값을 찾는 것이 좋다고 생각합니다. 그렇게 되는 경우에는 Dao와 DB는 데이터를 저장, 삭제만 하게 되고, DB에서 찾은 값을 적절한 도메인으로 바꾸는 역할은 하지 않게 됩니다. DB는 단순 저장의 목적으로만 사용한다고 생각해야 하는게 맞을까요?

혹은 Dao에 move라는 메서드를 만들어 Board와 동일한 로직(기물을 찾고 기물이 존재하면 새로운 값으로 수정하고, 없으면 추가하도록)을 수행하도록 하는 방법도 생각해봤습니다.

앤지의 조언 부탁드립니다!

soominsohn replied:
하루의 의견에 동의합니다!

도메인 내부인 Board에서 Position과 Piece를 Map으로 가지고 있기 때문에,
출발지점, 도착지점의 기물을 찾을 때 DB에 접근하지 않고 Board의 map을 사용하고 움직이고 난 뒤의 결과를 다시 DB에 저장하는게 좋아 보입니다.
또한 하루의 질문처럼 DB는 단순 저장의 목적으로 사용되므로, Dao에서 move 처럼 로직이 섞인 메서드는 도메인의 책임을 침범하므로 지양하는게 좋습니다.

학습하시면서 늘 좋은 결론에 도달하시는군요. 👍

soominsohn commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -13,17 +12,13 @@ public Piece(Camp camp) {
     public abstract void validateRouteObstacles(Set<Piece> piecesOnRoute);
 
-    public abstract Set<Point> findRoute(Point from, Point to);
+    public abstract Set<Point> findRoute(MoveType moveType, Point from, Point to);
```
> ### 2️⃣`validateRouteObstacles()`, `findRoute()` 메서드
> 
> 현재는 부모 클래스에 경로 탐색을 위한 `validateRouteObstacles()`, `findRoute()` 메서드가 정의되어 있는데,
> 
> `General`, `Guard`, `Soldier`와 같이 한 칸만 움직이는 기물은 실제로 경로가 존재하지 않아서 빈 구현으로 남겨두고 있습니다.
> 
> 구현하지 않은 메서드가 있는건 좋은 설계인 것 같지 않아서 고민이 됩니다!
> 
> `Routable` 같은 인터페이스를 따로 분리해 경로 개념이 필요한 기물만 해당 기능을 구현하도록 하는 방향이 더 괜찮을까요?
> 
> 이 구조를 더 명확하게 만들 수 있는 조언 주시면 감사하겠습니다!

하루의 의견에 완전 찬성입니다!
General, Guard, Soldier는 validateRouteObstacles(), findRoute()가 의미 없는데도,
부모 클래스의 메서드 때문에 빈 구현을 하고 있어야 하는건 불필요해보여요. Piece에는 공통으로 필요한 메서드만 있으면 된다고 생각이들고, 경로 탐색이 필요한 기물만 `Routable` 을 구현하는게 더 객체지향적인 코드인 것 같아요! 반영 기다리고 있겠습니다!

eunseongu replied:
Routable 인터페이스를 만들어 경로가 존재하는 기물만 validateRouteObstacles, findRoute를 구현하도록 했습니다.

하지만 Board에서는 Piece 타입으로만 기물들을 다루고 있어서
인터페이스 메서드를 호출하려면 다운캐스팅이 필요해지는 상황이 발생했고,
이를 위해 getType() 등으로 타입을 확인하는 구조는 적절하지 않다고 판단했습니다.

그래서 현재는 Piece에서 해당 메서드의 기본 구현으로 예외를 던지도록 하고,
경로가 존재하는 기물만 이를 오버라이딩하여 사용하도록 변경했습니다.

더 좋은 방법이 있을까요?
조언 부탁드립니다! 😊

soominsohn replied:
Routable을 컴포지션으로 활용하는 방법도 있을 것 같아요.

- Routable을 인터페이스로 선언한 뒤 Piece의 필드로 가지고 있고 Routable인터페이스에 validateRouteObstacles, findRoute 메서드 선언
- 경로를 검증해야는 기물들에는 DefaultRoutable이라는 구현체를 만들어 메서드를 구현
- 경로 검증이 필요없는 기물들은 NoRoutable 구현체를 만들어 메서드에서 예외를 던져주는 방식

쓰다 보니 이렇게 사용하려면 네이밍도 조금 바꿔야할 것 같고, 과연 이렇게까지 구현해야할까라는 생각이 들긴 합니다. 
지금 방식인 Piece에서 해당 메서드의 기본 구현으로 예외던지기도 괜찮아 보여요. 😄
하루가 생각하기에 적절한 방법을 사용하면 좋을 것 같습니다.

soominsohn commented on src/main/java/janggi/board/PalaceArea.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Set;
+
+public class PalaceArea {
```
> ### 1️⃣궁성의 좌표 관리
> 
> board에서 궁성의 영역에 들어서면 PalaceArea에서 관련 로직을 처리하도록 시도해보았습니다.
> 
> 하지만 그렇게 하다 보니 move, validateRoute 등 Board와 PalaceArea 간에 중복되는 로직이 생겼습니다.
> 
> 그래서 현재는 PalaceArea가 자신의 좌표들만 관리하도록 하여 궁성에 해당하는 위치인지 여부를 판단하는 책임만 가지도록 했습니다.
> 
> 다만 그러다보니 PalaceArea의 역할이 적은 것 같다는 생각이 들었습니다.
> 
> 하지만 “궁성에 속한 좌표는 궁성 객체가 아는 것이 자연스럽다”고 생각해 현재 구조를 유지했습니다.
> 
> 이에 대한 앤지의 생각은 어떠신가요?
> 

네! 코드를 작성하시다보면 역할이 적은 객체를 작성할 때도 있는데요, 역할이 적다고 해서 그 객체가 하는 동작을 다른 객체를 옮길 필요는 없을 것 같아요. 
PalaceArea는 역할 분리라는 관점에서 잘 생성하신 것 같아 지금 하루의 구조에 찬성합니다! ㅎㅎ

soominsohn reviewed (changes requested):
안녕하세요 하루~~
2단계 구현도 잘 해주셨군요!
코멘트 몇 개 남겼으니 확인 부탁드려요~~

eunseongu committed:
refactor: 불필요한 변수 제거 및 주석 추가

eunseongu committed:
refactor: 테스트에서 PieceDao를 검증하도록 변경

- TestPieceDao를 제거하고 DBConfig를 외부에서 주입받도록 변경합니다.

eunseongu committed:
refactor: 패키지 분리

eunseongu committed:
refactor: 도메인이 DAO를 의존하지 않도록 변경

eunseongu committed:
refactor: 데이터베이스에서 기물을 찾을 때 null을 반환하지 않도록 변경

eunseongu committed:
refactor: 데이터베이스에서 승리한 진영을 찾을 때 null을 반환하지 않도록 변경

eunseongu committed:
refactor: GameDao 리팩토링

- try-catch 이후 값을 반환하지 않고 예외를 던지도록 변경합니다.

eunseongu committed:
refactor: Piece의 경로 메서드에서 예외를 던지도록 변경

- 경로가 존재하지 않는 기물을 고려하여 Piece의 기본 구현에서 예외를 던지도록 변경합니다.
- Guard, General, Soldier 기물에서 불필요한 메서드 구현을 제거합니다.

eunseongu committed:
refactor: 올바른 인덱스를 사용하도록 변경

soominsohn commented on src/main/java/janggi/view/View.java:
```diff
@@ -52,59 +49,84 @@ private boolean parseYesOrNo(String response) {
-        if (placedPieces.get(new Point(j, i)) == null) {
+    private void displayPiece(PieceDao pieceDao, int i, int j) {
+        Point point = new Point(j, i);
+        Optional<Piece> piece = pieceDao.findByPoint(point);
```
View에서 PieceDao를 알고 있어도 될까요~?
View가 아닌 다른 곳에서 게임에 사용되고 있는 position과 piece를 가져와서 view에게 넘겨주면 어때요?

soominsohn reviewed (approved):
안녕하세요 하루~
리팩토링 해주신 부분 잘 봤습니다. 
주신 질문들은 답변 달아놨어요! 저도 많이 배웠답니다. 👍 
레벨1 수고많으셨고, 레벨2도 파이팅하세요!! 👋
